Add developer documentation (#4)
julea-io/haura: docs/src/betree/design.md

    "cb6" [label="CB" shape="box" style="filled" fillcolor="#33333333"]
    "cb7" [label="CB" shape="box" style="filled" fillcolor="#33333333"]
    "cb8" [label="CB" shape="box" style="filled" fillcolor="#33333333"]
    "cb9" [label="CB" shape="box" style="filled" fillcolor="#33333333"]
    
    "l1" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    "l2" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    "l3" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    "l4" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    "l6" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    "l7" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    "l8" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    "l9" [label="Leaf" style="filled" fillcolor="#0000FF55"]
    
    "root" -> "c1"
    "root" -> "c2"
    
    "c1" -> "in1"
    "c2" -> "in2"
    
    "in1" -> "cb1"
    "in1" -> "cb2"
    "in1" -> "cb3"
    "in1" -> "cb4"
    "in2" -> "cb6"
    "in2" -> "cb7"
    "in2" -> "cb8"
    "in2" -> "cb9"
    
    "cb1" -> "l1"
    "cb2" -> "l2"
    "cb3" -> "l3"
    "cb4" -> "l4"
    "cb6" -> "l6"
    "cb7" -> "l7"
    "cb8" -> "l8"
    "cb9" -> "l9"
}
```

> The `Dataset` interacts mainly with the actual Bε-tree, which

Author: Yeah we should use B-epsilon everywhere, it's just less of hassle and easier to read and write anyway.

    "root" -> "ds1" [style="dashed" label="Object Pointer"]
    "root" -> "ds2" [style="dashed" label="Object Pointer"]
    "root" -> "ds3" [style="dashed" label="Object Pointer"]
    "ds1" -> "DMU"
    "ds2" -> "DMU"
    "ds3" -> "DMU"
    
    {rank=same; "root"; "ds1"}
    {rank=same; "root"; "ds2"}
    {rank=same; "root"; "ds3"}
}
```

### Data Management

On-disk operations and storage allocation are handled by the Data Management
layer. This layer also implements the copy-on-write semantics required for
snapshots, done in delayed deallocation and accounting of a dead-list of blocks.

The `Handler` manages the actual bitmap handling for all allocations and
deallocations and is also responsible for tracking the number of blocks
distributed (Space Accounting).

To keep track of specific locations of allocated blocks, or free ranges of
blocks rather, bitmaps are used.  Wrapped around `SegmentAllocator`s, these can
be used to allocate block ranges at any position in a specific `SegmentId` or
request specific allocations at given offsets.

`SegementId`s refer to 1 GiB large ranges of blocks on a storage tier, though
the Id is unique over all storage tiers.

#### Copy on Write

> The Data Management Layer is also responsible to handle copy on write

Reviewer: ```suggestion
The Data Management Layer is also responsible to handle copy-on-write
```
(consistency)

preservation.  This is handle by checking if any snapshots of the dataset
contain the afflicted node (via `Generation`), if this is the case the
`dead_list` contained in the root tree is updated, to contain the storage
location of the old version of the node on the next sync.

### Storage Pool

> As the abstraction over specific hardware types and raid configurations the data

Reviewer: ```suggestion
As the abstraction over specific hardware types and RAID configurations the data
```

> management unit interacts for all IO operation with the storage pool layer.

Reviewer: ```suggestion
management unit interacts for all I/O operation with the storage pool layer.
```
